Documentation/design-docs/native-hosting.md

```diff
@@ -0,0 +1,176 @@
+Apps using the component hosting functionality would ship this library as part ot the app. So similarly to `apphost`/`comhost`/`ijwhost` the `nethost` will be shipped as part of .NET Core SDK to be included in customer's projects.
+
+This library would expose two APIs
+
+### Load managed component and get a function pointer
+``` C++
+int nethost_load_assembly_method(
+        const char * assembly_path,
+        const char * type_name,
+        const char * method_name,
+        const void * reserved,
+        void ** delegate);
+)
+```
+This API will
+* Locate the assembly using the `assembly_path` and its `.runtimeconfig.json` and determine the frameworks it requires to run. (Note that only framework dependent components will be supported for now).
+* If the process doesn't have CoreCLR loaded (more specifically `hostpolicy` library)
+  * Using the `.runtimeconfig.json` resolve required frameworks (just like if running an application) and load the runtime.
+* Else the CoreCLR is already loaded, in that case validate that required frameworks for the component can be satisfied by the runtime.
+  * If the required frameworks are not already present, fail. No support to load additional frameworks for now.
+* Call into the runtime (`System.Private.CoreLib` specifically)
+  * Create a new isolated `AssemblyLoadContext` (possibly reusing for the same components)
+  * Load the component's assembly into it
+  * Find the requested `type_name` and `method_name`
+  * Return a native callable function pointer to the requested method
+
+The `reserved` argument is currently not used and must be set to `nullptr`.
+
+If the runtime is initialized by this function, it will only be populated with framework assemblies (its TPA), none of the component's assemblies will be loaded into the default context.
+
+*As proposed there would be no support for unloading components. For discussion on possible solutions see open issues below.*
+
+### Locate `hostfxr`
+``` C++
+typedef void(* nethost_get_hosfxr_path_result_fn)(const char * hostfxr_path);
+
+int nethost_get_hostfxr_path(
+        nethost_get_hostfxr_path_result_fn result,
+        const char * assembly_path); // Optional
+```
+
+This API locates the `hostfxr` and returns its path by calling the `result` function. (This callback design is chosen so that it's clear and easy to define memory ownership.)
+
+`assembly_path` is optional:
+* If not specified, the "muxer" (`dotnet.exe`) behavior is used to locate the `hostfxr`
+* If specified, the `apphost` behavior is used
+
+
+## Improve API to run application
+New API will be added to `hostfxr`.
+
+``` C++
+struct hostfxr_runtime_properties
+{
+    int count;
+    const char ** keys;
+    const char ** values;
+};
+
+struct hostfxr_parameters
+{
+    int size;
+    const char * host_path;
+    const char * dotnet_root;
+    const char * app_path;
+    hostfxr_runtime_properties additional_properties;
+};
+
+int hostfxr_main_with_parameters(
```

Comment:
Considering that we are redesigning startup API is it worth to accommodate https://github.com/dotnet/core-setup/issues/4596 too?

So we get something like:

``` C
// hostfxr_initialize reads and resolves deps file 

hostfxr_initialize(&hostfxr_runtime);

// hostfxr_runtime can expose resolved dll locations, native directories etc.

// setup hostfxr_parameters

// hostfxr_main_with_parameters reuses cached dependencies resolved in hostfxr_initialize

hostfxr_main_with_parameters(hostfxr_runtime, hostfxr_parameters);
```

Reply:
I would not say we're redesigning it... we're trying to add support for additional properties and the current API is not extensible.
The proposed API on the other hand is, so if we wanted to do something like you described we in theory could do that in the future without introducing a new API.

I actually considered doing something similar - an API which would initialize the runtime (all the way up to loading and initializing CoreCLR) and then return, so that the native host could make direct calls on the `coreclr`. It would work for "run the app" scenarios, but it's really tricky to make it behave reasonable for the "load managed component" scenarios.

Given that this would be basically a "low-level" API, I would like it to work in all the scenarios - so far I was not able to come up with a good solution.

* If you're interested the biggest problem is what happens if there already is a runtime in the process. Currently coreclr doesn't lend itself to a reuse, we can make it work in specific scenarios, but not really for the general case of letting people call what they want. *

Another potential problem with your proposed design is that the `hostfxr_runtime` would contain information from both `hostfxr` and `hostpolicy`, so it would have to somehow store state for both libraries. Not impossible, but could be tricky given that these two libraries can have very different versions (`hostfxr` is always latest, so for example 3.0, `hostpolicy` can be any version, even `1.1`).  Which basically means we could only support this new API on apps targeting 3.0 and above, or it would have to include some kind of "fallback to old behavior" mode.

Reply:
> Given that this would be basically a "low-level" API, I would like it to work in all the scenarios - so far I was not able to come up with a good solution.

`hostfxr_main_with_parameters` doesn't address all scenarios either it doesn't mean we can't address more issues in "run the app" case.

Reply:
@vitek-karas IT would be ideal if we could take this into consideration, today we do a double resolve and that could be completely removed if we could preserve the state we as part of the resolve phase (essentially the processed deps file)

Reply:
The latest version of this design should allow for both of these use cases:
* Ability to modify properties before initializing the runtime
* Ability to read properties computed by the hosting layer before the runtime is initialized - this way you can get the native search paths
* Doing the above in such a way that the native app controls the code flow, so it's possible to read property, load another library, pass control over to that library (along with the context handle) set additional properties and start the runtime&app from the other library.